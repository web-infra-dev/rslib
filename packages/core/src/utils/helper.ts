import fs from 'node:fs';
import fsP from 'node:fs/promises';
import path from 'node:path';
import color from 'picocolors';
import type { PkgJson } from '../types';
import { logger } from './logger';

/**
 * Node.js built-in modules.
 * Copied from https://github.com/webpack/webpack/blob/dd44b206a9c50f4b4cb4d134e1a0bd0387b159a3/lib/node/NodeTargetPlugin.js#L12-L72
 */
export const nodeBuiltInModules: Array<string | RegExp> = [
  'assert',
  'assert/strict',
  'async_hooks',
  'buffer',
  'child_process',
  'cluster',
  'console',
  'constants',
  'crypto',
  'dgram',
  'diagnostics_channel',
  'dns',
  'dns/promises',
  'domain',
  'events',
  'fs',
  'fs/promises',
  'http',
  'http2',
  'https',
  'inspector',
  'inspector/promises',
  'module',
  'net',
  'os',
  'path',
  'path/posix',
  'path/win32',
  'perf_hooks',
  'process',
  'punycode',
  'querystring',
  'readline',
  'readline/promises',
  'repl',
  'stream',
  'stream/consumers',
  'stream/promises',
  'stream/web',
  'string_decoder',
  'sys',
  'timers',
  'timers/promises',
  'tls',
  'trace_events',
  'tty',
  'url',
  'util',
  'util/types',
  'v8',
  'vm',
  'wasi',
  'worker_threads',
  'zlib',
  /^node:/,

  // cspell:word pnpapi
  // Yarn PnP adds pnpapi as "builtin"
  'pnpapi',
];

export async function calcLongestCommonPath(
  absPaths: string[],
): Promise<string | null> {
  if (absPaths.length === 0) {
    return null;
  }

  // we support two cases
  // 1. /packages-a/src/index.ts
  // 2. D:/packages-a/src/index.ts
  const sep = path.posix.sep as '/';

  const splitPaths = absPaths.map((p) => p.split(sep));
  let lcaFragments = splitPaths[0]!;
  for (let i = 1; i < splitPaths.length; i++) {
    const currentPath = splitPaths[i]!;
    const minLength = Math.min(lcaFragments.length, currentPath.length);

    let j = 0;
    while (j < minLength && lcaFragments[j] === currentPath[j]) {
      j++;
    }

    lcaFragments = lcaFragments.slice(0, j);
  }

  let lca = lcaFragments.length > 0 ? lcaFragments.join(sep) : sep;

  const stats = await fsP.stat(lca);
  if (stats?.isFile()) {
    lca = path.dirname(lca);
  }

  return lca;
}

export const readPackageJson = (rootPath: string): undefined | PkgJson => {
  const pkgJsonPath = path.join(rootPath, './package.json');

  if (!fs.existsSync(pkgJsonPath)) {
    logger.warn(`package.json does not exist in the ${rootPath} directory`);
    return;
  }

  try {
    return JSON.parse(fs.readFileSync(pkgJsonPath, 'utf8'));
  } catch (err) {
    logger.warn(`Failed to parse ${pkgJsonPath}, it might not be valid JSON`);
    return;
  }
};

export const isObject = (obj: unknown): obj is Record<string, any> =>
  Object.prototype.toString.call(obj) === '[object Object]';

export const isEmptyObject = (obj: object): boolean => {
  return Object.keys(obj).length === 0;
};

export function pick<T, U extends keyof T>(
  obj: T,
  keys: ReadonlyArray<U>,
): Pick<T, U> {
  return keys.reduce(
    (ret, key) => {
      if (obj[key] !== undefined) {
        ret[key] = obj[key];
      }
      return ret;
    },
    {} as Pick<T, U>,
  );
}

export function omit<T extends object, U extends keyof T>(
  obj: T,
  keys: ReadonlyArray<U>,
): Omit<T, U> {
  return Object.keys(obj).reduce(
    (ret, key) => {
      if (!keys.includes(key as U)) {
        ret[key as keyof Omit<T, U>] = obj[key as keyof Omit<T, U>];
      }
      return ret;
    },
    {} as Omit<T, U>,
  );
}

export { color };
