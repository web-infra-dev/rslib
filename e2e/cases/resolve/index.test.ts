import { join } from 'node:path';
import { buildAndGetResults } from '@e2e/helper';
import { expect, test } from 'vitest';

test('resolve data url', async () => {
  const fixturePath = join(__dirname, 'data-url');
  const { entries, isSuccess } = await buildAndGetResults(fixturePath);

  expect(isSuccess).toBeTruthy();
  expect(entries.esm).toMatchInlineSnapshot(`
    "// CONCATENATED MODULE: data:text/javascript,export default 42
    const javascript_export_default_42 = 42;
    // CONCATENATED MODULE: ./index.ts
    console.log('x:', javascript_export_default_42);
    "
  `);
});

// TODO: false module path is different from linux and windows
// EXTERNAL MODULE: <ROOT>/rslib/e2e/cases/resolve/false/./browser-false/util (ignored)
test.todo('resolve false', async () => {
  const fixturePath = join(__dirname, 'false');
  const { entries, isSuccess } = await buildAndGetResults(fixturePath);

  expect(isSuccess).toBeTruthy();
  expect(entries.esm).toMatchSnapshot();
});

test('resolve node protocol', async () => {
  const fixturePath = join(__dirname, 'node-protocol');
  const { entries, isSuccess } = await buildAndGetResults(fixturePath);

  expect(isSuccess).toBeTruthy();
  expect(entries.esm).toMatchInlineSnapshot(`
    "import * as __WEBPACK_EXTERNAL_MODULE_node_path__ from \\"node:path\\";
    // CONCATENATED MODULE: external \\"node:path\\"
    var external_node_path_namespaceObject = __WEBPACK_EXTERNAL_MODULE_node_path__;
    // CONCATENATED MODULE: ./index.ts
    const { join } = external_node_path_namespaceObject[\\"default\\"];
    export { join };
    "
  `);
});

test('resolve with condition exports', async () => {
  const fixturePath = join(__dirname, 'with-condition-exports');
  const { contents, isSuccess } = await buildAndGetResults(fixturePath);

  const nodeResults = Object.values(contents.esm0!);
  const browserResults = Object.values(contents.esm1!);

  expect(isSuccess).toBeTruthy();

  expect(nodeResults[0]).toContain('lib1 mjs');
  expect(nodeResults[1]).toContain('lib2 module');
  expect(nodeResults[2]).toContain('node');
  expect(nodeResults[3]).toContain('lib1 cjs');

  expect(browserResults[0]).toContain('lib1 mjs');
  expect(browserResults[1]).toContain('lib2 module');
  expect(browserResults[2]).toContain('browser');
  expect(browserResults[3]).toContain('lib1 cjs');
});

test('resolve with js extensions', async () => {
  const fixturePath = join(__dirname, 'with-js-extensions');
  const { entries, isSuccess } = await buildAndGetResults(fixturePath);

  expect(isSuccess).toBeTruthy();
  expect(entries.esm).toMatchInlineSnapshot(`
    "// CONCATENATED MODULE: ./example.mts
    const value = 1;
    // CONCATENATED MODULE: ./index.ts
    console.log(value);
    "
  `);
});

test('resolve with main fields', async () => {
  const fixturePath = join(__dirname, 'with-main-fields');
  const { contents, isSuccess } = await buildAndGetResults(fixturePath);
  const results = Object.values(contents);

  expect(isSuccess).toBeTruthy();
  expect(Object.values(results[0]!)[0]).toMatchInlineSnapshot(`
    "// CONCATENATED MODULE: ./node_modules/lib1/src/index.ts
    const value = 1;
    // CONCATENATED MODULE: ./entry1.ts
    console.log(value);
    "
  `);
  expect(Object.values(results[1]!)[0]).toContain('main');
  expect(Object.values(results[2]!)[0]).toContain('browser');
});
