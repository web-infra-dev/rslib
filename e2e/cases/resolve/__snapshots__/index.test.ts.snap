// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`resolve false 1`] = `
"var __webpack_modules__ = {
    \\"?69d8\\": function() {
    /* (ignored) */ }
};
/************************************************************************/ // The module cache
var __webpack_module_cache__ = {};
// The require function
function __webpack_require__(moduleId) {
    // Check if module is in cache
    var cachedModule = __webpack_module_cache__[moduleId];
    if (void 0 !== cachedModule) return cachedModule.exports;
    // Create a new module (and put it into the cache)
    var module = __webpack_module_cache__[moduleId] = {
        exports: {}
    };
    // Execute the module function
    __webpack_modules__[moduleId](module, module.exports, __webpack_require__);
    // Return the exports of the module
    return module.exports;
}
/************************************************************************/ // webpack/runtime/compat_get_default_export
(()=>{
    // getDefaultExport function for compatibility with non-harmony modules
    __webpack_require__.n = function(module) {
        var getter = module && module.__esModule ? function() {
            return module['default'];
        } : function() {
            return module;
        };
        __webpack_require__.d(getter, {
            a: getter
        });
        return getter;
    };
})();
// webpack/runtime/define_property_getters
(()=>{
    __webpack_require__.d = function(exports, definition) {
        for(var key in definition)if (__webpack_require__.o(definition, key) && !__webpack_require__.o(exports, key)) Object.defineProperty(exports, key, {
            enumerable: true,
            get: definition[key]
        });
    };
})();
// webpack/runtime/has_own_property
(()=>{
    __webpack_require__.o = function(obj, prop) {
        return Object.prototype.hasOwnProperty.call(obj, prop);
    };
})();
/************************************************************************/ // CONCATENATED MODULE: ./bar.ts
const bar = 'bar';
// EXTERNAL MODULE: <ROOT>/e2e/cases/resolve/false/./browser-false/util (ignored)
var util_ignored_ = __webpack_require__(\\"?69d8\\");
var util_ignored_default = /*#__PURE__*/ __webpack_require__.n(util_ignored_);
console.log('foo:', util_ignored_default());
// the value of \`foo\` should \`{}\`
console.log('bar: ', bar);
"
`;
